### PR TITLE
feat(trace): add withdraw trie root

### PR DESCRIPTION
**Pull request by 0xmountaintop**
reference:
+ https://github.com/ethereum-optimism/optimism/blob/develop/l2geth/core/state_transition.go#L129-L133
+ https://github.com/ethereum-optimism/optimism/blob/develop/l2geth/rollup/fees/rollup_fee.go#L171
+ https://github.com/scroll-tech/scroll/blob/feat/new_bridge_contracts/contracts/src/L2/predeploys/L2MessageQueue.sol

---
## for `FeeVaultAddress`
@Thegaram do you think we should hardcode `FeeVaultAddress` in rollup/rcfg/config.go (this is also how OP does with their predeploys as in https://github.com/ethereum-optimism/optimism/blob/develop/l2geth/rollup/rcfg/config.go)

Or should we just provide it in `ChainCfg` (current approach)?

### CODE REVIEW

**Review by colinlyguo — DISMISSED** [review]

**Review by colinlyguo — DISMISSED** [review]

**Review by colinlyguo — DISMISSED** [review]
LGTM

**Comment by Thegaram** [comment]
> @Thegaram do you think we should hardcode `FeeVaultAddress` in rollup/rcfg/config.go (this is also how OP does with their predeploys as in https://github.com/ethereum-optimism/optimism/blob/develop/l2geth/rollup/rcfg/config.go)
> 
> Or should we just provide it in `ChainCfg` (current approach)?

For now I'd prefer keeping it as it is, we can always change in the future. (It's a backward compatible change if the fee vault address does not change.)

**Review by icemelon — DISMISSED** [review]

File: rollup/rcfg/config.go [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import (
+	"math/big"
+
+	"github.com/scroll-tech/go-ethereum/common"
+)
+
+// TODO:
+// 1. use config
+// 2. allow different networks / hardforks
+// 3. vefify in consensus layer when decentralizing sequencer
```
Comment (icemelon) [inline rollup/rcfg/config.go]:
```suggestion
// 3. verify in consensus layer when decentralizing sequencer
```

File: rollup/rcfg/config.go [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+var (
+	// L2MessageQueueAddress is the address of the L2MessageQueue
+	// predeploy
+	// see scroll-tech/scroll/contracts/src/L2/predeploys/L2MessageQueue.sol
```
Comment (icemelon) [inline rollup/rcfg/config.go]:
```suggestion
	// see contracts/src/L2/predeploys/L2MessageQueue.sol
```

File: rollup/withdrawtrie/withdraw_trie.go [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+)
+
+// StateDB represents the StateDB interface
+// required to get withdraw trie root
+type StateDB interface {
+	GetState(common.Address, common.Hash) common.Hash
+}
+
+// ReadWTRSlot reads WithdrawTrieRoot slot in L2MessageQueue predeploy, i.e., `messageRoot`
+// in scroll-tech/scroll/contracts/src/libraries/common/AppendOnlyMerkleTree.sol
```
Comment (icemelon) [inline rollup/withdrawtrie/withdraw_trie.go]:
```suggestion
// in contracts/src/libraries/common/AppendOnlyMerkleTree.sol
```

Comment (0xmountaintop) [reply rollup/withdrawtrie/withdraw_trie.go]: fixed in https://github.com/scroll-tech/go-ethereum/pull/217/commits/40668d88ae01833e0eadd5e183aa6875598f410b

File: rollup/rcfg/config.go (hunk 2 above)
Comment (0xmountaintop) [reply rollup/rcfg/config.go]: fixed in https://github.com/scroll-tech/go-ethereum/pull/217/commits/40668d88ae01833e0eadd5e183aa6875598f410b

File: rollup/rcfg/config.go (hunk 1 above)
Comment (0xmountaintop) [reply rollup/rcfg/config.go]: fixed in https://github.com/scroll-tech/go-ethereum/pull/217/commits/40668d88ae01833e0eadd5e183aa6875598f410b

**Pull request merged**